pkg/tekton/pipelines.go

```diff
@@ -261,6 +261,16 @@ func ApplyPipelineRun(tektonClient tektonclient.Interface, ns string, run *v1alp
 	resourceName := run.Name
 	resourceInterface := tektonClient.TektonV1alpha1().PipelineRuns(ns)
 
+	f := func() error {
+		_, err := tektonClient.TektonV1alpha1().Pipelines(ns).Get(run.Spec.PipelineRef.Name, metav1.GetOptions{})
+		return err
+	}
+	// Make sure the Pipeline has been created and registered before we try to create the PipelineRun.
+	err := util.RetryWithInitialDelay(2*time.Millisecond, 30*time.Second, f)
```

Comment:
That said, I am a bit sceptical about the `CreateOrUpdatePipeline` implementation:

```golang
	answer, err := resourceInterface.Create(created)
	if err == nil {
		return answer, nil
	}
```

The assumption seems to be that if there is an error it must be that the pipeline existed. That's not the case. There could be many different errors. What's missing here is something like:

```
	if errors.IsAlreadyExists(err) {
		// let try an update
	}
```

Reply:
So my theory, which is probably wrong, was that maybe the `Pipeline` isn’t getting returned by the API yet when the Tekton reconciler responds to the creation of the `PipelineRun`. I’m now pretty sure that’s not the case, and that the issue is a stale cache in the Tekton reconciler that needs to be addressed there. I’m going to remove this change.

Reply:
>  I’m now pretty sure that’s not the case, and that the issue is a stale cache in the Tekton reconciler that needs to be addressed there. I’m going to remove this change.

that would make more sense. From the perspective of this code, the pipeline exists, otherwise you would not end up in  in this piece of code. That said, the error check we are doing on Create is wrong. If you are going to revert this change, could you please open a new issue to address this?

Reply:
fwiw, we don't want to update if the `Pipeline` already exists - that means we've got a problem somewhere, like a reused build number, so erroring out is what we want.